nanofi/src/api/ecu.c

```diff
@@ -0,0 +1,494 @@
+
+    if (errno != 0) {
+        printf("Invalid interval value specified\n");
+        return -1;
+    }
+
+    errno = 0;
+    *port_num = (uint64_t)(strtoul(params->tcp_port, NULL, 10));
+    if (errno != 0) {
+        printf("Cannot convert tcp port to numeric value\n");
+        return -1;
+    }
+    return 0;
+}
+
+void setup_signal_action() {
+    struct sigaction action;
+    memset(&action, 0, sizeof(sigaction));
+    action.sa_handler = signal_handler;
+    sigaction(SIGTERM, &action, NULL);
+    sigaction(SIGINT, &action, NULL);
+}
+
+nifi_proc_params setup_nifi_processor(tailfile_input_params * input_params, const char * processor_name, void(*callback)(processor_session *, processor_context *)) {
+    nifi_proc_params params;
+    nifi_port port;
+    port.port_id = input_params->nifi_port_uuid;
+
+    nifi_instance * instance = create_instance(input_params->instance, &port);
+    add_custom_processor(processor_name, callback);
+    standalone_processor * proc = create_processor(processor_name, instance);
+    params.instance = instance;
+    params.processor = proc;
+    return params;
+}
+
+void add_to_hash_table(flow_file_record * ffr, uint64_t offset, char * uuid) {
+    struct processor_params * pp = NULL;
+    HASH_FIND_STR(procparams, uuid, pp);
+    if (pp == NULL) {
+        pp = (struct processor_params*)malloc(sizeof(struct processor_params));
+        memset(pp, 0, sizeof(struct processor_params));
+        char * uuid_str = (char *)malloc((strlen(uuid) + 1) * sizeof(char));
+        strcpy(uuid_str, uuid);
+        pp->uuid_str = uuid_str;
+        HASH_ADD_KEYPTR(hh, procparams, pp->uuid_str, strlen(pp->uuid_str), pp);
+    }
+    add_flow_file_record(&pp->ff_list, ffr);
+    pp->curr_offset = offset;
+}
+
+void delete_all_flow_files_from_proc(const char * uuid) {
+    struct processor_params * pp = NULL;
+    HASH_FIND_STR(procparams, uuid, pp);
+    if (pp) {
+        struct flow_file_list * head = pp->ff_list;
+        while (head) {
+            struct flow_file_list * tmp = head;
+            free_flowfile(tmp->ff_record);
+            head = head->next;
+            free(tmp);
+        }
+        pp->ff_list = head;
+    }
+}
+
+void delete_completed_flow_files_from_proc(const char * uuid) {
+    struct processor_params * pp = NULL;
+    HASH_FIND_STR(procparams, uuid, pp);
+    if (pp) {
+        struct flow_file_list * head = pp->ff_list;
+        while (head) {
+            struct flow_file_list * tmp = head;
+            if (tmp->complete) {
+                free_flowfile(tmp->ff_record);
+                head = head->next;
+                free(tmp);
+            }
+            else {
+                break;
+            }
+        }
+        pp->ff_list = head;
+    }
+}
+
+uint64_t get_current_offset(char * uuid) {
+    struct processor_params * pp = NULL;
+    HASH_FIND_STR(procparams, uuid, pp);
+    if (pp) {
+        return pp->curr_offset;
+    }
+    return 0;
```

Comment:
Isn't 0 a vlaid offset?

Reply:
Yes it is a valid offset. 0 is returned the very first time the processor is invoked and we do not find the processor in the hashtable.

Reply:
In subsequent calls, the processor is found in the hashtable.